### ISSUE TITLE
Proposal: switch ignore case

### ISSUE THREAD

**Issue by lechu445** [dotnet/csharplang#1321]
Currently switch statement on strings is case-sensitive. 
Consider a way to provide an ignore-case switch construct so it could be optimized by compiler.

My example proposal:
```
switch(text, StringComparison.OrdinalIgnoreCase)
{
    case "some_value1":  
        break;  
    case "some_value2":  
        break;  
}
```

**Comment by yaakov-h** [comment]
How would this work?

**Comment by tannergooding** [comment]
Why not just `switch(text.ToLower())` or `switch(text.ToLowerInvariant())`?

**Comment by tannergooding** [comment]
It would seem like a more generally useful feature (but still one that was just syntax sugar) would be to allow:

`switch (T, IComparable)` and have it lowered to an `if...else` chain that checks `T` against `case` using the specified comparer.

**Comment by yaakov-h** [comment]
@tannergooding How about `IComparer` or `IEqualityComparer`, too?

**Comment by mburbea** [comment]
@tannergooding , string switches generate a dictionary last I checked, so it should be possible to change the dictionary to have a `StringComparer` passed to it.

Also, if you're going to do the normalization approach, you should do `text.ToUpperInvariant()` based on the guidance found here. https://msdn.microsoft.com/en-us/library/bb386042.aspx

**Comment by Richiban** [comment]
Active patterns would solve this issue, although the syntax might look a little different, e.g.

```csharp
switch(s)
{
    case IgnoreCase("some_value1"):
        break;  
    case IgnoreCase("some_value2"):
        break;  
}
```

**Comment by Richiban** [comment]
I would prefer to leave the `switch(x, y)` syntax for switching on two values at the same time (i.e. switching on a ValueTuple):

```csharp

public static string FizzBuzz(int n)
{
    switch(i % 3, i % 5)
    {
        case (0, 0): return "fizzbuzz";
        case (0, _): return "fizz";
        case (_, 0): return "buzz";
        default:     return n.ToString();
    }
}

```

**Comment by vladd** [comment]
@tannergooding 
"Why not just `switch(text.ToLower())`" — because this is [not the same](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEAMABLgIwDcAsAFC4DMxATIQMLUDe1hXx9pAbMQAsxEgA4AFAEpO3DlW4KRRKCUIBeQgCIAkpsrzFXUsqYbNgBkA9Mw6QCc464e7GAdAFEAjuggAbKOJUkQigGIIBlGAQAS0wAc2YAezAABwhoqATMFwB5BAATGN9tWMwEhDhmaDhJSX0nEXtHetdPbz8AkhcAFQSAGQSAd0QpIJDuvsHhyXDImPik1PTMnPzCnxq67gBfai2gA=).

**Comment by gafter** [comment]
`string.ToLower()` uses the invariant culture's mapping, but that is an old incorrect frozen-in-time Unicode case-lower mapping. See https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/CaseInsensitiveComparison.cs for the lower-case mapping used by the VB compiler. It has a special case for the Turkish `İ` to compensate for the possibility that the platform lacks a suitable modern Unicode implementation (for most purposes removing the special case would have no effect).

**Comment by tannergooding** [comment]
@vladd, right. Looks like `ToUpperInvariant()` is the correct version to use:
* @mburbea's comment https://github.com/dotnet/csharplang/issues/1321#issuecomment-365415546
* The actual `String.Equals` implementation for `OrdinalIgnoreCase` does something similar
  * https://source.dot.net/#System.Private.CoreLib/src/System/String.Comparison.cs,949

In either case, I would still think supporting a more general syntax would be desirable:
* `switch (T, IComparable)`
* `switch (T, IComparer)`
* `switch (T, IEqualityComparer)`

That makes it more generally useful and still handles the desired case.

**Comment by HaloFour** [comment]
@tannergooding 

For simple `switch` statements that does make sense, but in my opinion it makes a lot less sense now that `switch` is effectively a vehicle for pattern matching.  It's not clear or obvious where any comparer would apply to anything other than constant patterns, and there might be some assumption that it would apply to recursive patterns, e.g.:

```cs
switch (person, StringComparer.OrdinalIgnoreCase) {
    case { LastName is "Gates" }: 
        // matches when LastName is "gates" or "GATES"?
        break;
}
```

**Comment by gafter** [comment]
`ToUpperInvariant` is not correct for two reasons. First, it uses the buggy Unicode table of the invariant locale. Second, it performs an upper-case mapping instead of a lower-case mapping. The advice of [CA1308](https://msdn.microsoft.com/en-us/library/bb386042.aspx) is founded on the incorrect logic that a lower-case mapping produces results that do not round-trip. But that is true of the upper-case mapping as well - the whole point of the mapping is to map *different* characters to *the same* one for the purposes of considering them to be "the same". The Unicode specifications, and the historical VB language implementations, base case folding (which is used to compare strings ignoring case) on a locale-independent lower-case mapping.

**Comment by lechu445** [comment]
Dictionary can use `IEqualityComparer<TKey>` in constructor  https://source.dot.net/#System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs,69, so I would think adding `switch (T, IEqualityComparer<T>)` overload.

`switch(text.ToUpper())`, `swich(text.ToUpperInvariant())`, etc. - this approach requires extra string conversion and allocation.

**Comment by TheConstructor** [comment]
Any status on this?

If `string`-matching already uses a `Dictionary` to implement `switch`, I would also support a syntax to specify the `IEqualityComparer<T>`.
- It may not be combinable with `case`-patterns, but wouldn't block you from adding `when`. There just needs to be a decision, if patterns can be used, ignoring the setting, or would be syntax-errors, if a comparer is specified.
- It offers a natural way to adjust switch-behavior for other types
- Lastly in most cases, where I want to use `StringComparer.OrdinalIgnoreCase`, I usually have `string`-constant cases

For me it would seem to a good addition to the language
